tsdb/chunks/chunks.go

```diff
@@ -241,45 +246,63 @@ func MergeOverlappingChunks(chks []Meta) ([]Meta, error) {
 	}
 	newChks := make([]Meta, 0, len(chks)) // Will contain the merged chunks.
 	newChks = append(newChks, chks[0])
-	last := 0
 	for _, c := range chks[1:] {
 		// We need to check only the last chunk in newChks.
-		// Reason: (1) newChks[last-1].MaxTime < newChks[last].MinTime (non overlapping)
-		//         (2) As chks are sorted w.r.t. MinTime, newChks[last].MinTime < c.MinTime.
-		// So never overlaps with newChks[last-1] or anything before that.
-		if c.MinTime > newChks[last].MaxTime {
+		// Reason: (1) newChks[len(newChks)-2].MaxTime < newChks[len(newChks)-1].MinTime (non overlapping)
+		//         (2) As chks are sorted w.r.t. MinTime, newChks[len(newChks)-1].MinTime < c.MinTime.
+		// So never overlaps with newChks[len(newChks)-1-1] or anything before that.
+		if c.MinTime > newChks[len(newChks)-1].MaxTime {
 			newChks = append(newChks, c)
-			last++
 			continue
 		}
-		nc := &newChks[last]
-		if c.MaxTime > nc.MaxTime {
-			nc.MaxTime = c.MaxTime
-		}
-		chk, err := MergeChunks(nc.Chunk, c.Chunk)
+		dividedChks, err := MergeChunks(newChks[len(newChks)-1].Chunk, c.Chunk, MaxSamplesInChunk)
 		if err != nil {
 			return nil, err
 		}
-		nc.Chunk = chk
+		newChks = append(newChks[:len(newChks)-1], dividedChks...)
+		// Appending the first len(newChks)-2 chunks with the divided chunks.
+		// The len(newChks)-1 th chunk would have been merged and hence is overwritten.
 	}
-
 	return newChks, nil
 }
 
 // MergeChunks vertically merges a and b, i.e., if there is any sample
 // with same timestamp in both a and b, the sample in a is discarded.
-func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
+// Also breaks chunks that are larger than a particular limit.
+func MergeChunks(a, b chunkenc.Chunk, limit int) ([]Meta, error) {
+	var newMetaArray []Meta
 	newChunk := chunkenc.NewXORChunk()
 	app, err := newChunk.Appender()
 	if err != nil {
 		return nil, err
 	}
 	ait := a.Iterator(nil)
 	bit := b.Iterator(nil)
+	numSamples := 0
+	var mint, maxt int64
+	first := true // flag needed to check if looking at the first sample of the chunk and subsequenty set mint
```

Comment:
I think we don't need this variable as well. What about doing setting `mint` to `math.MinMaxInt64` and relying on that?

Reply:
I thought I would be explicit with the `first` variable. But Okay!